cpp/arcticdb/storage/azure/azure_storage.hpp

```diff
@@ -0,0 +1,127 @@
+
+    AzureStorage(const LibraryPath &lib, OpenMode mode, const Config &conf);
+
+    /**
+     * Full object path in Azure bucket.
+     */
+    std::string get_key_path(const VariantKey& key) const;
+
+  protected:
+    void do_write(Composite<KeySegmentPair>&& kvs);
+
+    void do_update(Composite<KeySegmentPair>&& kvs, UpdateOpts opts);
+
+    template<class Visitor>
+    void do_read(Composite<VariantKey>&& ks, Visitor &&visitor, ReadKeyOpts opts);
+
+    void do_remove(Composite<VariantKey>&& ks, RemoveOpts opts);
+
+    template<class Visitor>
+    void do_iterate_type(KeyType key_type, Visitor &&visitor, const std::string &prefix);
+
+    bool do_key_exists(const VariantKey& key);
+
+    bool do_supports_prefix_matching() {
+        return true;
+    }
+
+    bool do_fast_delete() {
+        return false;
+    }
+
+  private:
+    std::string blob_container_url_;
+    Azure::Storage::Blobs::BlobContainerClient container_client_;
+    std::string root_folder_;
+    bool connect_to_azurite_;
+    unsigned int request_timeout_;
+    Azure::Storage::Blobs::UploadBlockBlobFromOptions upload_option_;
+    Azure::Storage::Blobs::DownloadBlobToOptions download_option_;
+};
+
+
+class AzureStorageFactory final : public StorageFactory<AzureStorageFactory> {
+    using Parent = StorageFactory<AzureStorageFactory>;
+    friend Parent;
+
+  public:
+    using Config = arcticdb::proto::azure_storage::Config;
+    using StorageType = AzureStorage;
+
+    AzureStorageFactory(const Config &conf) :
+        conf_(conf) {
+    }
+  private:
+    auto do_create_storage(const LibraryPath &lib, OpenMode mode) {
+            return AzureStorage(lib, mode, conf_);
+    }
+
+    Config conf_;
+};
+
+inline arcticdb::proto::storage::VariantStorage pack_config(const std::string &container_name) {
+    arcticdb::proto::storage::VariantStorage output;
+    arcticdb::proto::azure_storage::Config cfg;
+    cfg.set_container_name(container_name);
+    util::pack_to_any(cfg, *output.mutable_config());
+    return output;
+}
+
+inline arcticdb::proto::storage::VariantStorage pack_config(
+        const std::string &container_name,
+        const std::string &credential_name,
+        const std::string &credential_key,
+        const std::string &endpoint
+        ) {
+    arcticdb::proto::storage::VariantStorage output;
+    arcticdb::proto::azure_storage::Config cfg;
+    cfg.set_container_name(container_name);
+    cfg.set_credential_name(credential_name);
+    cfg.set_credential_key(credential_key);
+    cfg.set_endpoint(endpoint);
+    util::pack_to_any(cfg, *output.mutable_config());
+    return output;
+}
+
+template<typename ConfigType>
+std::shared_ptr<Azure::Storage::StorageSharedKeyCredential> get_azure_credentials(const ConfigType& conf) {
+    return std::make_shared<Azure::Storage::StorageSharedKeyCredential>(conf.credential_name(), conf.credential_key());
+}
+
+} //namespace arcticdb::azure
+
+#define ARCTICDB_AZURE_STORAGE_H_
```

Comment:
Is this ever used?

Reply:
Used in `azure_storage.hpp`, like other storage adapters